### PR TITLE
[SPARK-47541][SQL][FOLLOWUP] Fix `AnsiTypeCoercion` to handle ArrayType

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
Fixing array_join.

### Why are the changes needed?
ANSI tests failing.

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
CollationSuite in both ANSI and non-ansi mode.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by mihailom-db** [comment]
Fix for https://github.com/apache/spark/pull/45693

**Comment by mihailom-db** [comment]
@dongjoon-hyun Could you merge this in to fix the broken tests in master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 1]
```diff
@@ -422,7 +422,8 @@ abstract class TypeCoercionBase {
           case None => c
         }
 
-      case aj @ ArrayJoin(arr, d, nr) if !ArrayType(StringType).acceptsType(arr.dataType) &&
+      case aj @ ArrayJoin(arr, d, nr)
+        if !AbstractArrayType(StringTypeAnyCollation).acceptsType(arr.dataType) &&
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:426]: nit. We need two more spaces for indentation.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs). Thank you, @mihailom-db .

I verified ANSI mode manually too.

```
$ SPARK_ANSI_SQL_MODE=true build/sbt "sql/testOnly *CollationSuite -- -z complex"
...
[info] All tests passed.
[success] Total time: 134 s (02:14), completed Apr 11, 2024, 3:07:39 PM
```

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0.

**Pull request closed without merging**